rust/routee-compass-powertrain/src/model/prediction/prediction_model_record.rs

```diff
@@ -21,130 +16,105 @@ pub struct PredictionModelRecord {
                     *underlying_model.clone(),
-                    config.name.clone(),
-                    config.speed_unit,
-                    (*speed_lower_bound, *speed_upper_bound),
-                    *speed_bin_size,
-                    config.grade_unit,
-                    (*grade_lower_bound, *grade_upper_bound),
-                    *grade_bin_size,
+                    config.input_features.clone(),
+                    feature_bounds.clone(),
                     config.energy_rate_unit,
                 )?;
                 Arc::new(model)
             }
         };
         let ideal_energy_rate = prediction_model_ops::find_min_energy_rate(
             &prediction_model,
-            config.speed_unit,
-            config.grade_unit,
+            &config.input_features,
             &config.energy_rate_unit,
         )?;
 
         let real_world_energy_adjustment = config.real_world_energy_adjustment.unwrap_or(1.0);
-        let cache_policy = match &config.float_cache_policy {
-            Some(cache_config) => Some(FloatCachePolicy::from_config(cache_config.clone())?),
-            None => None,
-        };
 
         Ok(PredictionModelRecord {
             name: config.name.clone(),
             prediction_model,
             model_type: config.model_type.clone(),
-            speed_unit: config.speed_unit,
-            grade_unit: config.grade_unit,
+            input_features: config.input_features.clone(),
+            distance_unit: config.distance_unit,
             energy_rate_unit: config.energy_rate_unit,
             ideal_energy_rate,
             real_world_energy_adjustment,
-            cache: cache_policy,
         })
     }
 }
 
 impl PredictionModelRecord {
     pub fn predict(
         &self,
-        speed: (Speed, &SpeedUnit),
-        grade: (Grade, &GradeUnit),
-        distance: (Distance, &DistanceUnit),
+        state: &mut [StateVariable],
+        state_model: &StateModel,
     ) -> Result<(Energy, EnergyUnit), TraversalModelError> {
-        let (distance, distance_unit) = distance;
-        let (speed, speed_unit) = speed;
-        let (grade, grade_unit) = grade;
-
-        let energy_rate = match &self.cache {
-            Some(cache) => {
-                let key = vec![speed.as_f64(), grade.as_f64()];
-                match cache.get(&key)? {
-                    Some(er) => EnergyRate::from(er),
-                    None => {
-                        let (energy_rate, _energy_rate_unit) = self
-                            .prediction_model
-                            .predict((speed, speed_unit), (grade, grade_unit))?;
-                        cache.update(&key, energy_rate.as_f64())?;
-                        energy_rate
+        let (distance, distance_unit) =
+            state_model.get_distance(state, fieldname::EDGE_DISTANCE, Some(&self.distance_unit))?;
+        let mut feature_vector: Vec<f64> = Vec::new();
+        for (name, input_feature) in &self.input_features {
```

Comment:
here we start to step through the declared set of input features set for this prediction model record. from your PR description, these changes
> [update] our vehicle models to accept N input features rather than hard coding them as speed and grade model

but right here, we restrict the possible types of input feature units to Speed and Grade. so, in a sense, this implementation really only supports those two units. i'm guessing to some degree, you're leaving the door open here to add Temperature and Mass in the future, and they should be updated here when that is the case. however, is there a reason to block `Time`, `Distance` and `Custom` `InputFeature` types here?

Reply:
I've added the ability to use more of the input features but right now if you add a model in that uses other features it will still block you when it tries to find [the minimum energy rate](https://github.com/NREL/routee-compass/blob/bbe6b7c2e3fbd0ce64a5be4468ed49c72852c55d/rust/routee-compass-powertrain/src/model/prediction/prediction_model_ops.rs#L54). There's not a clear path forward to generate sample values for something like Time, Distance and especially a custom variable and so maybe this opens the door to future work to allow the user to specify the sample values (or ranges) for each of the input features.